File: go/ppc/rule.go
package capitan

import (
    "errors"
    "strings"
)

type Ast interface{}

type StringAst struct {
    Value string
}

type TemplateAst struct {
    Templates []string
}

type EqualAst struct {
    Left, Right Ast
}

type NotEqualAst struct {
    Left, Right Ast
}

type RewriteAst struct {
    Source      Ast
    Replacement [2]Ast
}

type MatchAst struct {
    Source   Ast
    Patterns [][2][]Ast
}

type TrueAst struct{}
type FalseAst struct{}

type OptionAst struct {
    Name string
}

type Rule []Ast

type Options struct {
    Nested bool
    Strict bool
}

func parseAst(tokens []string) (Ast, []string, error) {
    if len(tokens) == 0 {
        return nil, nil, errors.New("Unexpected end of input")
    }

    switch tokens[0] {
    case "true":
        return TrueAst{}, tokens[1:], nil
    case "false":
        return FalseAst{}, tokens[1:], nil
    case "option":
        if len(tokens) < 2 {
            return nil, nil, errors.New("Option name expected")
        }
        return OptionAst{Name: tokens[1]}, tokens[2:], nil
    case "=":
        if len(tokens) < 4 {
            return nil, nil, errors.New("Invalid equal expression")
        }
        lhs, _, _ := parseAst(tokens[1:2])
        rhs, rest, _ := parseAst(tokens[3:])
        return EqualAst{Left: lhs, Right: rhs}, rest, nil
    case "!=":
        if len(tokens) < 3 {
            return nil, nil, errors.New("Invalid not equal expression")
        }
        lhs, _, _ := parseAst(tokens[1:2])
        rhs, rest, _ := parseAst(tokens[2:])
        return NotEqualAst{Left: lhs, Right: rhs}, rest, nil
    case "->":
        if len(tokens) < 4 {
            return nil, nil, errors.New("Invalid rewrite expression")
        }
        lhs, _, _ := parseAst(tokens[1:2])
        rhs, rest, _ := parseAst(tokens[3:])
        return RewriteAst{Source: lhs, Replacement: [2]Ast{rhs, rhs}}, rest, nil
    default:
        return StringAst{Value: tokens[0]}, tokens[1:], nil
    }
}

func parseRule(tokens []string) (Rule, error) {
    var rule Rule
    for len(tokens) > 0 {
        if tokens[0] == "," {
            tokens = tokens[1:]
            continue
        }
        ast, rest, err := parseAst(tokens)
        if err != nil {
            return nil, err
        }
        rule = append(rule, ast)
        tokens = rest
    }
    return rule, nil
}

func Create(ruleText string) (Rule, error) {
    tokens := strings.Fields(ruleText)
    if len(tokens) == 0 || tokens[0] != "where" {
        return nil, errors.New("Rule must start with 'where'")
    }
    return parseRule(tokens[1:])
}

func (r Rule) Options() Options {
    opts := Options{false, false}
    for _, ast := range r {
        if option, ok := ast.(OptionAst); ok {
            switch option.Name {
            case "nested":
                opts.Nested = true
            case "strict":
                opts.Strict = true
            }
        }
    }
    return opts
}

func (r Rule) IsStrict() bool {
    return r.Options().Strict
}

File: python/ppc/rule.py
from dataclasses import dataclass
from typing import List, Tuple, Union, Optional

@dataclass
class String:
    value: str

@dataclass
class Template:
    templates: List[str]

@dataclass
class Equal:
    left: 'Ast'
    right: 'Ast'

@dataclass
class NotEqual:
    left: 'Ast'
    right: 'Ast'

@dataclass
class Rewrite:
    source: 'Ast'
    replacement: Tuple['Ast', 'Ast']

@dataclass
class Match:
    source: 'Ast'
    patterns: List[Tuple['Ast', List['Ast']]]

class True_:
    pass

class False_:
    pass

@dataclass
class Option:
    name: str

Ast = Union[String, Template, Equal, NotEqual, Rewrite, Match, True_, False_, Option]
Rule = List[Ast]

@dataclass
class Options:
    nested: bool = False
    strict: bool = False

class ParseError(Exception):
    pass

def parse_ast(tokens: List[str]) -> Tuple[Ast, List[str]]:
    if not tokens:
        raise ParseError("Unexpected end of input")

    if tokens[0] == "true":
        return True_(), tokens[1:]
    elif tokens[0] == "false":
        return False_(), tokens[1:]
    elif tokens[0] == "option":
        if len(tokens) < 2:
            raise ParseError("Option name expected")
        return Option(tokens[1]), tokens[2:]
    elif tokens[0] == "=":
        if len(tokens) < 4:
            raise ParseError("Invalid equal expression")
        lhs, _ = parse_ast(tokens[1:2])
        rhs, rest = parse_ast(tokens[3:])
        return Equal(lhs, rhs), rest
    elif tokens[0] == "!=":
        if len(tokens) < 3:
            raise ParseError("Invalid not equal expression")
        lhs, _ = parse_ast(tokens[1:2])
        rhs, rest = parse_ast(tokens[2:])
        return NotEqual(lhs, rhs), rest
    elif tokens[0] == "->":
        if len(tokens) < 4:
            raise ParseError("Invalid rewrite expression")
        lhs, _ = parse_ast(tokens[1:2])
        rhs, rest = parse_ast(tokens[3:])
        return Rewrite(lhs, (rhs, rhs)), rest
    else:
        return String(tokens[0]), tokens[1:]

def parse_rule(tokens: List[str]) -> Rule:
    rule = []
    while tokens:
        if tokens[0] == ",":
            tokens = tokens[1:]
            continue
        ast, tokens = parse_ast(tokens)
        rule.append(ast)
    return rule

def create(rule_text: str) -> Optional[Rule]:
    tokens = rule_text.split()
    if not tokens or tokens[0] != "where":
        return None
    try:
        return parse_rule(tokens[1:])
    except ParseError:
        return None

def options(rule: Rule) -> Options:
    opts = Options()
    for ast in rule:
        if isinstance(ast, Option):
            if ast.name == "nested":
                opts.nested = True
            elif ast.name == "strict":
                opts.strict = True
    return opts

def is_strict(rule: Rule) -> bool:
    return options(rule).strict
